src/main.rs

```diff
@@ -156,21 +158,44 @@ impl Fixer {
     /// subsequently query. Return a description of the failing operation on
     /// error.
     fn build_file_info(file_name: &str) -> Result<FileInfo, String> {
-        // Get the debug session from file.
-        let msg = |op| format!("Unable to {} file {}", op, file_name);
-        let data = fs::read(file_name).map_err(|_| msg("read"))?;
+        let msg = |op: &str| format!("Unable to {} `{}`", op, file_name);
+
+        // Read the file.
+        let mut data = fs::read(file_name).map_err(|_| msg("read"))?;
+
+        // On some platforms we have to get the debug info from another file.
+        // Get the name of that file, if there is one.
+        let file_name2 = match Object::peek(&data) {
+            FileFormat::Pe => {
+                let pe_object = Object::parse(&data).map_err(|_| msg("parse"))?;
+                if let Object::Pe(pe) = pe_object {
+                    // PE files should contain a pointer to a PDB file.
+                    let pdb_file_name = pe.debug_file_name().ok_or_else(|| msg("find PDB for"))?;
+                    Some(pdb_file_name.to_string())
+                } else {
+                    panic!(); // Impossible: peek() said it was a PE object.
+                }
+            }
+            _ => None,
+        };
+        if let Some(file_name2) = file_name2 {
```

Comment:
Move this to a separate function, so that you  can write something more natural like:
```
if let Some(file_name2) = MaybeGetExternalDebugInfo() {
  [...]
}
```

Reply:
I tried that but it didn't feel good. I had to duplicate the `msg` closure, and the return type of the new function was `Result<Option<String>, String>` which is pretty gross.

Instead I split off the second half of this function into `FileInfo::new`. That keeps all the fallible I/O operations in this function, while moving the infallible `FileInfo` construction into a `FileInfo` method. That's a much nicer way to break this function into smaller pieces.